### PR TITLE
net: lib: wifi_credentials: Add new functions and params

### PR DESCRIPTION
**Pull request by simensrostad**
Add function to clear all stored credentials and to check if there are any valid entries in storage.

Add storing of channel.

Make wifi exts auto connect feature pass in preferred channel to the wifi stack.

### CODE REVIEW

**Review by ioannisg — COMMENTED** [review]
Looks good to me.
Just a question @simensrostad : shouldn't this API extension compiled-in only if non-STATIC (or not BACKEND_NONE) configuration is selected? Ignore comment if this is already the case. (if not, I believe it is cleaner and possibly more memory-optimized if STATIC credential option does not include this api extension) Thoughts?

**Comment by simensrostad** [comment]
> Looks good to me. Just a question @simensrostad : shouldn't this API extension compiled-in only if non-STATIC (or not BACKEND_NONE) configuration is selected? Ignore comment if this is already the case. (if not, I believe it is cleaner and possibly more memory-optimized if STATIC credential option does not include this api extension) Thoughts?

Note sure of the general structure of the different backends and how each parameter is affected. But I assume that if this is the case, there are also other parameters that fall under this umbrella. Its only ssid and PSK that can be set with the static backend. (Kconfig) Its a separate effort to hide all these params.

cc @maxd-nordic can maybe elaborate here?

**Comment by simensrostad** [comment]
Extending the wifi cred shell module is missing

**Comment by maxd-nordic** [comment]
@simensrostad - @ioannisg is right, `wifi_credentials_is_empty` and `wifi_credentials_delete_all` only make sense if `CONFIG_WIFI_CREDENTIALS_BACKEND_NONE` is not set. (It doesn't make a huge difference though.)

**Comment by simensrostad** [comment]
> @simensrostad - @ioannisg is right, `wifi_credentials_is_empty` and `wifi_credentials_delete_all` only make sense if `CONFIG_WIFI_CREDENTIALS_BACKEND_NONE` is not set. (It doesn't make a huge difference though.)

Ah, right. I got mixed up with the params. Didn't consider the API 😢 Ill revisit this after vacation, I also need to update the tests. But I'm not a big fan of ifdeffing header files. An IS_ENABLED guard in the beginning of the function entries is perhaps cleaner and we avoid this. WDYT?

**Comment by ioannisg** [comment]
> > @simensrostad - @ioannisg is right, `wifi_credentials_is_empty` and `wifi_credentials_delete_all` only make sense if `CONFIG_WIFI_CREDENTIALS_BACKEND_NONE` is not set. (It doesn't make a huge difference though.)
> 
> Ah, right. I got mixed up with the params. Didn't consider the API 😢 Ill revisit this after vacation, I also need to update the tests. But I'm not a big fan of ifdeffing header files. An IS_ENABLED guard in the beginning of the function entries is perhaps cleaner and we avoid this. WDYT?

Thanks @simensrostad . Yeah, whatever solution results in compiling out unwanted code; up to you to decide... By the way, the same applies to other (already defined & implemented) APIs that make sense only if non-STATIC (or NONE) credentials' solution is selected. So you might want to go through the rest of the library and compile out a bunch of things :)

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 1]
```diff
@@ -191,7 +191,8 @@ int wifi_credentials_set_personal_struct(const struct wifi_credentials_personal
 
 int wifi_credentials_set_personal(const char *ssid, size_t ssid_len, enum wifi_security_type type,
 				  const uint8_t *bssid, size_t bssid_len, const char *password,
-				  size_t password_len, uint32_t flags)
+				  size_t password_len, uint32_t flags,
```
Comment (balaji-nordic) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:194]: I believe the new line is not needed.

File: subsys/net/lib/wifi_credentials/wifi_credentials.c [hunk 2]
```diff
@@ -353,4 +355,38 @@ void wifi_credentials_for_each_ssid(wifi_credentials_ssid_cb cb, void *cb_arg)
...
+{
+	int ret = 0;
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+	for (size_t i = 0; i < CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES; ++i) {
+		if (is_entry_used(i)) {
+			ret = wifi_credentials_delete_entry(i);
+			if (ret) {
+				LOG_ERR("Failed to delete WiFi credentials index %d, err: %d", i, ret);
+				goto exit;
```
Comment (balaji-nordic) [inline subsys/net/lib/wifi_credentials/wifi_credentials.c:381]: You can replace the `goto` statement a `break` .

**Review by balaji-nordic — CHANGES_REQUESTED** [review]

**Comment by simensrostad** [comment]
> > > @simensrostad - @ioannisg is right, `wifi_credentials_is_empty` and `wifi_credentials_delete_all` only make sense if `CONFIG_WIFI_CREDENTIALS_BACKEND_NONE` is not set. (It doesn't make a huge difference though.)
> > 
> > 
> > Ah, right. I got mixed up with the params. Didn't consider the API 😢 Ill revisit this after vacation, I also need to update the tests. But I'm not a big fan of ifdeffing header files. An IS_ENABLED guard in the beginning of the function entries is perhaps cleaner and we avoid this. WDYT?
> 
> Thanks @simensrostad . Yeah, whatever solution results in compiling out unwanted code; up to you to decide... By the way, the same applies to other (already defined & implemented) APIs that make sense only if non-STATIC (or NONE) credentials' solution is selected. So you might want to go through the rest of the library and compile out a bunch of things :)

In that case its reasonable to add these guards in a separate PR (Since its a broader effort). It would also do good to investigate if its needed at all, I.E the compiler already optimizes away the unneeded code.

JIRA:
https://nordicsemi.atlassian.net/browse/CIA-1190

**Review by divipillai — APPROVED** [review]

File: include/net/wifi_credentials.h [hunk 3]
```diff
@@ -94,6 +95,7 @@ struct wifi_credentials_enterprise {
  * @param[in] password_buf_len		length of password_buf
  * @param[out] password_len		length of password
  * @param[out] flags			flags
+ * @param[out] flags			channel
```
Comment (divipillai) [inline include/net/wifi_credentials.h:98]:
```suggestion
 * @param[out] channel			channel
```
I think this should be channel? fixes the build error.

Comment (simensrostad) [reply include/net/wifi_credentials.h]: thanks!

**Review by balaji-nordic — APPROVED** [review]
The commit message has a minor typo. `news` instead of `new`. Rest looks good

**Review by rlubos — APPROVED** [review]

**Pull request merged**